caas/ecs/application.go

```diff
@@ -0,0 +1,637 @@
+					},
+					{
+						Name:  aws.String("JUJU_K8S_APPLICATION_PASSWORD"),
+						Value: aws.String(config.IntroductionSecret),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_CONTROLLER_ADDRESSES"),
+						Value: aws.String(config.ControllerAddresses),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_CONTROLLER_CA_CERT"),
+						Value: aws.String(config.ControllerCertBundle),
+					},
+				},
+				MountPoints: []*ecs.MountPoint{
+					{
+						ContainerPath: aws.String(jujuDataDir),
+						SourceVolume:  aws.String("var-lib-juju"),
+					},
+					{
+						ContainerPath: aws.String("/charm/bin"),
+						SourceVolume:  aws.String("charm-data-bin"),
+					},
+					// DO we need this in init container?
+					// {
+					// 	ContainerPath: aws.String("/charm/containers"),
+					// 	SourceVolume:  aws.String("charm-data-containers"),
+					// },
+				},
+			},
+		},
+		Volumes: append(volumes, []*ecs.Volume{
+			// TODO: ensure no vol.Name conflict.
+			{
+				Name: aws.String("var-lib-juju"),
+				DockerVolumeConfiguration: &ecs.DockerVolumeConfiguration{
+					Scope:  aws.String("task"),
+					Driver: aws.String("local"),
+					Labels: a.labels(),
+				},
+			},
+			{
+				Name: aws.String("charm-data-bin"),
+				DockerVolumeConfiguration: &ecs.DockerVolumeConfiguration{
+					Scope:  aws.String("task"),
+					Driver: aws.String("local"),
+					Labels: a.labels(),
+				},
+			},
+		}...),
+	}
+	// container agent.
+	charmContainerDefinition := &ecs.ContainerDefinition{
+		Name:             aws.String("charm"),
+		Image:            aws.String(config.AgentImagePath),
+		WorkingDirectory: aws.String(jujuDataDir),
+		Cpu:              aws.Int64(10),
+		Memory:           aws.Int64(512),
+		DependsOn: []*ecs.ContainerDependency{
+			{
+				ContainerName: aws.String("charm-init"),
+				Condition:     aws.String("SUCCESS"),
+			},
+		},
+		Essential:  aws.Bool(true),
+		EntryPoint: strPtrSlice("/charm/bin/k8sagent"),
+		Command: strPtrSlice(
+			"unit",
+			"--data-dir", jujuDataDir,
+			"--charm-modified-version", strconv.Itoa(config.CharmModifiedVersion),
+			"--append-env",
+			"PATH=$PATH:/charm/bin",
+		),
+		// TODO: Health check/prob
+		Environment: []*ecs.KeyValuePair{
+			{
+				Name:  aws.String("JUJU_CONTAINER_NAMES"),
+				Value: aws.String(strings.Join(containerNames, ",")),
+			},
+			{
+				Name: aws.String(
+					"HTTP_PROBE_PORT", // constants.EnvAgentHTTPProbePort
+				),
+				Value: aws.String(
+					"3856", // constants.AgentHTTPProbePort
+				),
+			},
+		},
+		MountPoints: []*ecs.MountPoint{
+			{
+				ContainerPath: aws.String(jujuDataDir),
+				SourceVolume:  aws.String("var-lib-juju"),
+			},
+			{
+				ContainerPath: aws.String("/charm/bin"),
+				SourceVolume:  aws.String("charm-data-bin"),
+			},
+		},
+	}
+
+	for _, v := range containers {
+		// TODO: https://aws.amazon.com/blogs/compute/amazon-ecs-and-docker-volume-drivers-amazon-ebs/
+		// to use EBS volumes, it requires some docker storage plugin installed in the
```

Comment:
Are we going for EBS instead of instance storage? Is this so we can always re-attach storage (similar to k8s) when the tasks get rescheduled to other containers?

Reply:
Yes, We choose this as the default driver for EBS support.
- https://github.com/rexray/rexray